diff --git a/panda/board/drivers/can_common.h b/panda/board/drivers/can_common.h
index 11524dd1..a91b711d 100644
--- a/panda/board/drivers/can_common.h
+++ b/panda/board/drivers/can_common.h
@@ -7,7 +7,9 @@ uint32_t rx_buffer_overflow = 0;
 
 can_health_t can_health[PANDA_CAN_CNT] = {{0}, {0}, {0}};
 
-// Ignition detected from CAN meessages
+// Wake-up and ignition detected from CAN messages
+bool wake_on_can = false;
+uint32_t wake_on_can_cnt = 0U;
 bool ignition_can = false;
 uint32_t ignition_can_cnt = 0U;
 
@@ -158,6 +160,13 @@ void can_set_forwarding(uint8_t from, uint8_t to) {
 void ignition_can_hook(CANPacket_t *msg) {
   if (msg->bus == 0U) {
     int len = GET_LEN(msg);
+    const int TESLA_DI_GEAR_P = 1;
+    const int TESLA_DI_GEAR_R = 2;
+    const int TESLA_DI_GEAR_N = 3;
+    const int TESLA_DI_GEAR_D = 4;
+    static int tesla_gear = -1;
+    static bool tesla_seatbelt_latched = false;
+    static bool tesla_door_open = false;
 
     // GM exception
     if ((msg->addr == 0x1F1U) && (len == 8)) {
@@ -180,7 +189,7 @@ void ignition_can_hook(CANPacket_t *msg) {
       prev_counter_rivian = counter;
     }
 
-    // Tesla Model 3/Y exception
+    // Tesla Model 3/Y exception - wake while LV power state is active
     if ((msg->addr == 0x221U) && (len == 8)) {
       // 0x221 overlaps with Rivian which has random data on byte 0
       int counter = msg->data[6] >> 4;
@@ -189,12 +198,44 @@ void ignition_can_hook(CANPacket_t *msg) {
       if ((counter == ((prev_counter_tesla + 1) % 16)) && (prev_counter_tesla != -1)) {
         // VCFRONT_LVPowerState->VCFRONT_vehiclePowerState
         int power_state = (msg->data[0] >> 5U) & 0x3U;
-        ignition_can = power_state == 0x3;  // VEHICLE_POWER_STATE_DRIVE=3
-        ignition_can_cnt = 0U;
+        // ignition_can = power_state == 0x3;  // VEHICLE_POWER_STATE_DRIVE=3
+        // ignition_can_cnt = 0U;
+        wake_on_can = power_state != 0x0; // not VEHICLE_POWER_STATE_OFF
+        wake_on_can_cnt = 0U;
       }
       prev_counter_tesla = counter;
     }
 
+    // Tesla Model 3/Y exception - drive gears -> ignition
+    if ((msg->addr == 0x118U) && (len == 8)) {
+      int counter = msg->data[1] & 0x0FU;
+
+      static int prev_counter_tesla_gear = -1;
+      if ((counter == ((prev_counter_tesla_gear + 1) % 16)) && (prev_counter_tesla_gear != -1)) {
+        tesla_gear = (msg->data[2] >> 5) & 0x7;
+        if ((tesla_gear == TESLA_DI_GEAR_R) || (tesla_gear == TESLA_DI_GEAR_N) || (tesla_gear == TESLA_DI_GEAR_D)) {
+          ignition_can = true;
+        }
+        // don't drop ignition on SNA/INVALID gear values
+        if ((tesla_gear == TESLA_DI_GEAR_P) && (!tesla_seatbelt_latched || tesla_door_open)) {
+          ignition_can = false;
+        }
+        ignition_can_cnt = 0U;
+      }
+      prev_counter_tesla_gear = counter;
+    }
+
+    if ((msg->addr == 0x311U) && (len == 7)) {
+      int counter = msg->data[1] & 0x0FU;
+
+      static int prev_counter_tesla_ui = -1;
+      if ((counter == ((prev_counter_tesla_ui + 1) % 16)) && (prev_counter_tesla_ui != -1)) {
+        tesla_seatbelt_latched = ((msg->data[1] >> 5U) & 0x1U) != 0U;  // UI_warning->buckleStatus
+        tesla_door_open = ((msg->data[3] >> 4U) & 0x1U) != 0U;  // UI_warning->anyDoorOpen
+      }
+      prev_counter_tesla_ui = counter;
+    }
+
     // Mazda exception
     if ((msg->addr == 0x9EU) && (len == 8)) {
       ignition_can = (msg->data[0] >> 5) == 0x6U;
diff --git a/panda/board/drivers/drivers.h b/panda/board/drivers/drivers.h
index c6150670..98f94945 100644
--- a/panda/board/drivers/drivers.h
+++ b/panda/board/drivers/drivers.h
@@ -44,6 +44,8 @@ extern can_health_t can_health[PANDA_CAN_CNT];
 // Ignition detected from CAN meessages
 extern bool ignition_can;
 extern uint32_t ignition_can_cnt;
+extern bool wake_on_can;
+extern uint32_t wake_on_can_cnt;
 
 extern bool can_silent;
 extern bool can_loopback;
diff --git a/panda/board/main.c b/panda/board/main.c
index 2bce081e..a8c167da 100644
--- a/panda/board/main.c
+++ b/panda/board/main.c
@@ -173,7 +173,8 @@ static void tick_handler(void) {
 
       // tick drivers at 1Hz
       bool started = harness_check_ignition() || ignition_can;
-      bootkick_tick(started, recent_heartbeat);
+      bool wake_up = started || wake_on_can;
+      bootkick_tick(wake_up, recent_heartbeat);
 
       // increase heartbeat counter and cap it at the uint32 limit
       if (heartbeat_counter < UINT32_MAX) {
@@ -253,11 +254,15 @@ static void tick_handler(void) {
       if (ignition_can_cnt > 2U) {
         ignition_can = false;
       }
+      if (wake_on_can_cnt > 2U) {
+        wake_on_can = false;
+      }
 
       // on to the next one
       uptime_cnt += 1U;
       safety_mode_cnt += 1U;
       ignition_can_cnt += 1U;
+      wake_on_can_cnt += 1U;
 
       // synchronous safety check
       safety_tick(&current_safety_config);
